source/m2mnsdlinterface.cpp

```diff
@@ -1559,3 +1469,269 @@ nsdl_s * M2MNsdlInterface::get_nsdl_handle()
 {
     return _nsdl_handle;
 }
+
+void M2MNsdlInterface::handle_bootstrap_put_message(sn_coap_hdr_s *coap_header,
+                                                sn_nsdl_addr_s *address) {
+#ifndef M2M_CLIENT_DISABLE_BOOTSTRAP_FEATURE
+    tr_debug("M2MNsdlInterface::handle_bootstrap_message");
+    uint8_t response_code = COAP_MSG_CODE_RESPONSE_CHANGED;
+    sn_coap_hdr_s *coap_response = NULL;
+    bool success = false;
+    bool security_object = false;
+    uint16_t content_type = 0;
+
+    String resource_name = coap_to_string(coap_header->uri_path_ptr,
+                                          coap_header->uri_path_len);
+    tr_debug("M2MNsdlInterface::handle_bootstrap_message - uri %s", resource_name.c_str());
+
+    // Check incoming object
+    if (resource_name.compare(0,1,"0") == 0) {
+        security_object = true;
+        if (!_security) {
+            _security = new M2MSecurity(M2MSecurity::M2MServer);
+        }
+        if(_security) {
+            success = true;
+            _security->create_resource(M2MSecurity::ShortServerID, 1);
+            for (int i = 0; i <= M2MSecurity::ClientHoldOffTime; i++) {
+                M2MResource* res = _security->get_resource((M2MSecurity::SecurityResource)i);
+                if (res) {
+                    res->set_operation(M2MBase::PUT_ALLOWED);
+                }
+            }
+        }
+    }
+    else if (resource_name.compare(0,1,"1") == 0) {
+        success = true;
+    }
+
+    if (success) {
+        // Send delayed response if token is part of the message
+        if (coap_header->token_ptr) {
+            tr_debug("M2MNsdlInterface::handle_bootstrap_message - send delayed response");
+            coap_response = sn_nsdl_build_response(_nsdl_handle,
+                                                   coap_header,
+                                                   COAP_MSG_CODE_EMPTY);
+            if (coap_response) {
+                coap_response->msg_type = COAP_MSG_TYPE_ACKNOWLEDGEMENT;
+                sn_nsdl_send_coap_message(_nsdl_handle, address, coap_response);
+                sn_nsdl_release_allocated_coap_msg_mem(_nsdl_handle, coap_response);
+            }
+        }
+
+        if(coap_header->content_type_ptr) {
+            content_type = String::convert_array_to_integer(coap_header->content_type_ptr,
+                                                            coap_header->content_type_len);
+        }
+
+        tr_debug("M2MNsdlInterface::handle_bootstrap_message - content_type %d", content_type);
+        if (content_type != COAP_CONTENT_OMA_TLV_TYPE) {
+            success = false;
+        }
+        if (success) {
+            success = parse_bootstrap_message(coap_header, security_object);
+            // Set operations back to default ones
+            for (int i = 0; i <= M2MSecurity::ClientHoldOffTime; i++) {
```

Comment:
You can get all the resources using resources() API. Use that instead of this logic.